examples/asr/experimental/structured/speech_to_text_hybrid.py

```diff
@@ -0,0 +1,56 @@
+# Copyright (c) 2020, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import pytorch_lightning as pl
+
+from nemo.collections.asr.models import EncDecCTCModel, configs
+from nemo.core.config import hydra_runner
+from nemo.utils.config_utils import update_model_config
+from nemo.utils.exp_manager import exp_manager
+
+
+"""
+python speech_to_text_hybrid.py \
+    --config-path="conf" \
+    --config-name="quartznet_15x5" \
+    model.train_ds.manifest_filepath="/home/smajumdar/PycharmProjects/NeMo-som/examples/asr/an4/train_manifest.json" \
+    model.validation_ds.manifest_filepath="/home/smajumdar/PycharmProjects/NeMo-som/examples/asr/an4/test_manifest.json" \
+    trainer.gpus=1
+"""
+
+
+@hydra_runner(config_path="conf", config_name="quartznet_15x5")
+def main(cfg):
+    # Generate default asr model config
+    asr_model_config = configs.EncDecCTCModelConfig()
+
+    # Merge hydra updates with model config
+    # `drop_missing_subconfig=True` is necessary here. Without it, while the data class will instantiate and be added
+    # to the config, it contains test_ds.sample_rate = MISSING and test_ds.labels = MISSING.
+    # This will raise a OmegaConf MissingMandatoryValue error when processing the dataloaders inside
+    # model_utils.resolve_test_dataloaders(model=self) (used for multi data loader support).
+    # In general, any operation that tries to use a DictConfig with MISSING in it will fail,
+    # other than explicit update operations to change MISSING to some actual value.
+    asr_model_config = update_model_config(asr_model_config, cfg, drop_missing_subconfigs=True)
```

Comment:
Why is this not passed as part of hydra_runner()?
https://github.com/NVIDIA/NeMo/blob/main/nemo/core/config/set_config.py#L26-L28

hydra_runner() already accepts schema as a parameter  which is doing what you are doing here.

Reply:
Ok I have a concrete answer for this question now, now that I attempted to implement the above.
Here is the stack trace (truncated) when attempting the above 

```
Key 'betas' not in 'OptimConfig'
        full_key: model.optim.betas
        reference_type=Optional[OptimConfig]
        object_type=OptimConfig
```

This is one of the prime responsibilities of `update_model_config` - merging a `shell` optim + `shell` sched with a more concrete instantiation from yaml. Even with a perfectly specified Model config (say QN15x5) - if we tie one specific optimizer + scheduler config to it then it cannot be changed by the user - they cannot try another optimizer or another scheduler apart from the one specified by the config class.

The lines inside `update_model_config` which explicitly handle this are https://github.com/NVIDIA/NeMo/pull/1328/files#diff-b383ea91b2a34cbb3404da8fd4275a6ff5b036e0520a5b4491ec644b6e59c9b0R127-R170 

This chunk enables there to be a shell config class for optim and sched, and the actual details regarding any optim and sched inside the yaml. fyi @okuchaiev @ericharper @tkornuta-nvidia

Reply:
If you think hydra_runner() needs to be updated with your new `update_model_config` function, go right ahead.
My main concern is that we have have an API to do what you want to do, but instead of fixing the old API to work, you invented an entirely new API. Let's work to fix old API as opposed to reinventing the wheel.

Side note: I am certain that omegaconf supports what you are trying to do: https://omegaconf.readthedocs.io/en/latest/usage.html#merging-configurations. If it really isn't working, then our next steps should be to ensure that we are using hydra/omegaconf correctly inside of hydra_runner(). If we are using them correctly, and they still don't work, then we should file an issue on hydra/omegaconf in addition to updating hydra_runner() with `update_model_config`.

Reply:
@blisc Hydra runner was not meant for the purpose of merging Dataclass configs. It is fine iff your dataclass is perfectly matching your hydra overrides.

This is not the case for ModelPT configs. Think about the use case of optimizers alone. If you want to fix your optimizer to just Adam (not even AdamW) and not allow the user to change it at all, then its fine to use hydra as it was designed using the ConfigStore with a AdamConfig.

We dont want that. ModelPT config is generic to all Models - and needs to support optimizers and scheduler switching. Different optimizers can have different (disparate) arguments - SGD + momentum does not have betas. Adam doesnt have nesterov etc. So we need a base set of arguments (just the name and lr) inside the base config.

However this base config cannot be merged with an outer config - the keys betas, nesterov, momentum etc dont exist in the base config. I have tried it (and you can see the result above) and that is why I had to perform the merge operation in such a way.

The hydra_runner is not sufficient to work with this generic ModelPT config workflow at all and simply cannot support it. We can hack around it how much ever we want, but it wont support it cleanly. If you feel otherwise, please try to fix hydra runner to satisfy the requirement of a ModelPT level config that is generic enough to handle such cases.

> If it really isn't working, then our next steps should be to ensure that we are using hydra/omegaconf correctly inside of hydra_runner()

We are using private APIs of Hydra to support hydra_runner() - we are using it incorrectly by design in order to inject additional support for our cases, but this has limits. 

Finally, the only thing I can do to add this flexibility is to introduce this new API and simply call it inside hydra_runner(). But  TLT depends on this - and I do not intend to break the workflow that they have settled on. 

Note this is an experimental API - and is marked as such. You are more than welcome to attempt to modify hydra_runner to support this all on its own with just Hydra level support, and we can surely remove all of the `update_model_config` code at that point if it works.

Reply:
Also- the hydra_runner workflow (as we previously discussed) does not even work for non-script environments. So completely removing `update_model_config` is not an option.